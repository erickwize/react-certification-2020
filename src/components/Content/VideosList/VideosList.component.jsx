import React from 'react';
import MediaCard from '../MediaCard/MediaCard.component';

export default function VideosList(props) {
  if (!props.videos) {
    return 'Loading...';
  } else if (props.videos.length === 0) {
    return <div>No videos found</div>;
  } else {
    return (
      <>
        {props.videos.map((video) => (
          <MediaCard
            key={video.etag}
            title={video.title}
            imageUrl={video.imageUrl}
            body={video.description}
          />
        ))}
      </>
    );
  }
}
