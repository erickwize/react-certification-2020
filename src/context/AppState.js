/* global gapi*/
/* eslint-disable react/prop-types */
import React, { useReducer } from "react";
import AppContext from "./appContext";
import AppReducer from "./appReducer";
import validateItems from "../utils/validateItems";
import initGapi from "../utils/initGapi";
import { GET_RESULT_VIDEOS, GET_RELATED_VIDEOS, SET_LOADING } from "./types";

const AppState = (props) => {
  const initialState = {
    searchText: "Moonwalk",
    resultVideos: [],
    selectedVideo: {},
    relatedVideos: [],
    showPlayer: false,
    loading: true,
  };

  const [state, dispatch] = useReducer(AppReducer, initialState);

  // HANDLE INIT GAPI
  const handleInitGapi = async () => {
    initGapi();
    setTimeout(() => {
      getResultVideos(state.searchText);
    }, 1000);
  };

  // GET RESULT VIDEOS
  const getResultVideos = async (query) => {
    setLoading();
    let items;

    try {
      const response = await gapi.client.youtube.search.list({
        part: ["snippet"],
        maxResults: 50,
        q: query,
        type: ["video"],
      });
      items = validateItems(response.result.items);
    } catch (error) {
      console.error("getResultVideos: Something went wrong... ", error);
    }

    dispatch({
      type: GET_RESULT_VIDEOS,
      payload: { query, items },
    });
  };

  // GET RELATED VIDEOS
  const getRelatedVideos = async (video) => {
    setLoading();
    let items;

    try {
      const response = await gapi.client.youtube.search.list({
        part: ["snippet"],
        maxResults: 50,
        type: ["video"],
        relatedToVideoId: video.id,
      });
      items = validateItems(response.result.items);
    } catch (error) {
      console.error("getRelatedVideos: Something went wrong... ", error);
    }

    dispatch({
      type: GET_RELATED_VIDEOS,
      payload: { video, items },
    });
  };

  // SET LOADING
  const setLoading = () => {
    dispatch({ type: SET_LOADING });
  };

  return (
    <AppContext.Provider
      value={{
        searchText: state.searchText,
        resultVideos: state.resultVideos,
        selectedVideo: state.selectedVideo,
        relatedVideos: state.relatedVideos,
        showPlayer: state.showPlayer,
        loading: state.loading,
        handleInitGapi,
        getResultVideos,
        getRelatedVideos,
        setLoading,
      }}
    >
      {props.children}
    </AppContext.Provider>
  );
};

export default AppState;
