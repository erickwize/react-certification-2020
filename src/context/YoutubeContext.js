import React, { createContext, useState } from 'react'
import { saveArrayOnLocalStorage } from '../utils/localStorage'

const API_KEY = process.env.REACT_APP_YT_API_KEY
const MAX_RESULTS = 20

const YoutubeContext = createContext()

const SEARCH_URL = `https://www.googleapis.com/youtube/v3/search?part=snippet&key=${API_KEY}&type=video&maxResults=${MAX_RESULTS}&`

function YoutubeContextProvider({ children }) {
	const [videos, setVideos] = useState([])
	const defaultContext = {
		videos,
		search: term => {
			fetch(`${SEARCH_URL}q=${term}`)
				.then(response => response.json())
				.then(data => {
					if (data?.items?.length) {
						setVideos(data.items)
						saveArrayOnLocalStorage('videos', data.items)
					} else setVideos([])
				})
				.catch(error => {
					console.error(error)
				})
		},
		setVideosContext: newVideos => {
			setVideos(newVideos)
		},
	}

	return (
		<YoutubeContext.Provider value={defaultContext}>{children}</YoutubeContext.Provider>
	)
}

export { YoutubeContext, YoutubeContextProvider }
