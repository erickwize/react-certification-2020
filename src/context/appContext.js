import { createContext, useContext } from "react";

const appContext = createContext();

export const useAppContext = () => {
  const context = useContext(appContext);
  if (context === undefined) {
    throw new Error(
      "useAppContext must be used within AppState component tree"
    );
  }
  return context;
};

export default appContext;
