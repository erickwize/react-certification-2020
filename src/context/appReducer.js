import { GET_RESULT_VIDEOS, GET_RELATED_VIDEOS, SET_LOADING } from "./types";

export default (state, action) => {
  switch (action.type) {
    case GET_RESULT_VIDEOS:
      return {
        ...state,
        searchText: action.payload.query,
        resultVideos: action.payload.items,
        showPlayer: false,
        loading: false,
      };
    case GET_RELATED_VIDEOS:
      return {
        ...state,
        selectedVideo: action.payload.video,
        relatedVideos: action.payload.items,
        showPlayer: true,
        loading: false,
      };
    case SET_LOADING:
      return {
        ...state,
        loading: true,
      };
    default:
      return state;
  }
};
