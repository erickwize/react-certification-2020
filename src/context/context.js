// import Dexie from 'dexie'
import React, { useEffect, createContext, useReducer, memo } from 'react'
import { useFetch } from '../utils/hooks/useFetch'

export const Context = createContext(null)

export const ContextReducer = (state, action) => {
  switch (action.type) {
    case 'TOGGLE_THEME':
      return {
        ...state,
        theme: !state.theme,
      }
    case 'OPEN_MENU':
      return {
        ...state,
        menu: action.payload,
      }
    case 'GET_DATA':
      return {
        ...state,
        data: action.payload,
      }
    case 'SET_VIEW':
      return {
        ...state,
        view: action.payload,
      }
    case 'SET_TARGET':
      return {
        ...state,
        target: action.payload,
      }
    default:
      return state
  }
}

export const initialState = {
  target: 'Wizeline',
  theme: false,
  menu: false,
  view: 'home',
  data: [],
}
export const ContextProvider = ({ children }) => {
  const [state, dispatch] = useReducer(ContextReducer, initialState)
  const { data, isLoading } = useFetch(state.target)

  useEffect(() => {
    if (data) dispatch({ type: 'GET_DATA', payload: data })
  }, [data])

  return isLoading ? (
    <div> Loading...</div>
  ) : (
    <Context.Provider value={{ state, dispatch }}> {children} </Context.Provider>
  )
}
export default memo(ContextProvider)
