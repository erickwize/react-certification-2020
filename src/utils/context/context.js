import { createContext } from 'react';

export const Context = createContext({
  queryToSearch: null,
  setQueryFromSearchInput: null,
});
