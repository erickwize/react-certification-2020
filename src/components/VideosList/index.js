import React, { useContext } from 'react'
import styled from 'styled-components'
import { v4 as uuidv4 } from 'uuid'
import { YoutubeContext } from '../../context/YoutubeContext'
import VideoCard from '../VideoCard'

const VideosListStyled = styled.div`
	display: flex;
	flex-flow: wrap;
	justify-content: space-between;
`
const MessageStyled = styled.p`
	font-size: 18px;
`

export default function VideosList() {
	const { videos, loading, errorMsg } = useContext(YoutubeContext)
	return !loading ? (
		<>
			<VideosListStyled>
				{videos.map(item => (
					<VideoCard key={uuidv4()} item={item} />
				))}
				{videos.length === 0 && <MessageStyled>There&apos;s no results</MessageStyled>}
			</VideosListStyled>
			{errorMsg && <p>{errorMsg}</p>}
		</>
	) : (
		<p>Loading...</p>
	)
}
