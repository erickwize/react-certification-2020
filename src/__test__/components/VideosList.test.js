import { render } from '@testing-library/react'
import React from 'react'
import { YoutubeContextProvider } from '../../context/YoutubeContext'
import VideosList from '../../components/VideosList'

describe('VideosList testing', () => {
	it('VideosList Snapshot', () => {
		const { container } = render(
			<YoutubeContextProvider>
				<VideosList />
			</YoutubeContextProvider>
		)
		expect(container).toMatchSnapshot()
	})
})
