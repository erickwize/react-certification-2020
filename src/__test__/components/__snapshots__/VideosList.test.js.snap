// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VideosList testing VideosList Snapshot 1`] = `
<div>
  <div
    class="sc-jSgupP iCxkCv"
  >
    <p
      class="sc-gKsewC gPlKBg"
    >
      There's no results
    </p>
  </div>
  
</div>
`;
